### PR TITLE
feat: Update Private Endpoints implementation across the modules

### LINKED ISSUE #1190: [AVM Module Issue]: Private Endpoints Scope
### Check for previous/existing GitHub issues

- [X] I have checked for previous/existing GitHub issues

### Issue Type?

Feature Request

### Module Name

All modules supporting Private Endpoints.

### (Optional) Module Name if not listed above

_No response_

### (Optional) Module Version

_No response_

### Description

Private endpoints should support deployment in other scopes. For example, I'd like to deploy a storage account in one resource group but the private endpoint for that storage account in another.

This comes in handy when separating all networking components into one resource group in a subscription and have resources live in other resource groups.

This change will require an adjustment on all modules that create private endpoints.

### (Optional) Correlation Id

_No response_

### PR DESCRIPTION
**Pull request by krbar**
## Description

Updates to the implementation of private endpoint in multiple modules:
- Enabled custom resource group selection for Private Endpoints.
- Fixing outdated PE property `manualPrivateLinkServiceConnections`

> Note: the module `avm/res/search/search-service` which also required this update, has been updated via #1475 
> Several other modules have been excluded from this PR because they could not be merged at this time. Their PR implementation update will be tracked via #1583 

Resolves #1190 

## Pipeline Reference

| Pipeline |
| -------- |
| [avm.res.automation.automation-account](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.automation.automation-account.yml) |
| [avm.res.batch.batch-account](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.batch.batch-account.yml) |
| [avm.res.cache.redis](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.cache.redis.yml) |
| [avm.res.cognitive-services.account](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.cognitive-services.account.yml) |
| [avm.res.data-factory.factory](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.data-factory.factory.yml) |
| [avm.res.desktop-virtualization.host-pool](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.desktop-virtualization.host-pool.yml) |
| [avm.res.desktop-virtualization.workspace](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.desktop-virtualization.workspace.yml) |
| [avm.res.digital-twins.digital-twins-instance](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.digital-twins.digital-twins-instance.yml) |
| [avm.res.document-db.database-account](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.document-db.database-account.yml) |
| [avm.res.event-grid.domain](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.event-grid.domain.yml) |
| [avm.res.event-grid.namespace](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.event-grid.namespace.yml) |
| [avm.res.event-grid.topic](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.event-grid.topic.yml) |
| [avm.res.insights.private-link-scope](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.insights.private-link-scope.yml) |
| [avm.res.key-vault.vault](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.key-vault.vault.yml) |
| [avm.res.purview.account](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.purview.account.yml) |
| [avm.res.recovery-services.vault](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.recovery-services.vault.yml) |
| [avm.res.relay.namespace](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.relay.namespace.yml) |
| [avm.res.search.search-service](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.search.search-service.yml) |
| [avm.res.service-bus.namespace](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.service-bus.namespace.yml) |
| [avm.res.signal-r-service.signal-r](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.signal-r-service.signal-r.yml) |
| [avm.res.signal-r-service.web-pub-sub](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.signal-r-service.web-pub-sub.yml) |
| [avm.res.sql.server](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.sql.server.yml) |
| [avm.res.synapse.private-link-hub](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.synapse.private-link-hub.yml) |
| [avm.res.synapse.workspace](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.synapse.workspace.yml) |
| [avm.res.web.static-site](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.web.static-site.yml) |

## Type of Change

- [ ] Update to CI Environment or utlities (Non-module effecting changes)
- [x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [x] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

### CODE REVIEW

File: avm/res/storage/storage-account/README.md
Comment (AlexanderSehr) [inline avm/res/storage/storage-account/README.md:2055]: That's a bit of a suprising one we should double-check

File: avm/res/storage/storage-account/main.bicep
Comment (AlexanderSehr) [inline avm/res/storage/storage-account/main.bicep:72]: Intended removal?

Comment (AlexanderSehr) [reply avm/res/storage/storage-account/main.bicep:72]: I know there is a related issue open - yet it may have sneaked it's way into this PR.

File: avm/res/storage/storage-account/README.md
Comment (eriqua) [reply avm/res/storage/storage-account/README.md:2055]: @AlexanderSehr that happened in another recently merged PR for input parameters made null able, ref https://github.com/Azure/bicep-registry-modules/pull/1474/files#r1545613784

Comment (AlexanderSehr) [reply avm/res/storage/storage-account/README.md:2055]: Hey @eriqua, not sure I follow :o

File: avm/res/storage/storage-account/main.bicep
Comment (krbar) [reply avm/res/storage/storage-account/main.bicep:72]: > Intended removal?

Initially, it was not intended, however this default value causes a Bicep error and breaks the Set-AVMModule execution. And because the parameter description describes it as "Required", I assumed that there should no default here.

image

File: avm/res/batch/batch-account/main.bicep
Comment (eriqua) [inline avm/res/batch/batch-account/main.bicep:77]: I'd kindly ask to merge #1481 first, to simplify the review of this PR and correctly track the modules still missing v2 formatting

File: avm/res/app-configuration/configuration-store/README.md
Comment (eriqua) [inline avm/res/app-configuration/configuration-store/README.md:1]: I'd strongly suggest to not merge code bumping a module version if the corresponding pipeline is failing. Although unrelated, the failing pipeline will cause the new version to not be published.

File: avm/res/batch/batch-account/main.bicep
Comment (AlexanderSehr) [reply avm/res/batch/batch-account/main.bicep:77]: Hey @eriqua, I actually already reviewed this PR including the reformatting (which I tihnk applie to like 8-10 modules), wasn't too bad :)

File: avm/res/storage/storage-account/main.bicep
Comment (AlexanderSehr) [reply avm/res/storage/storage-account/main.bicep:72]: Fair, but it does change the intend. So what we should figure out (with @fblix) is how to handle it properly. Our options:
- Move the default away from the parameter to where it is used (like we did for diagnostics etc.)
  ```bicep
  networkAcls: !empty(networkAcls) ? {
      resourceAccessRules: networkAcls.?resourceAccessRules
      bypass: networkAcls.?bypass
      defaultAction: networkAcls.?defaultAction
      virtualNetworkRules: networkAcls.?virtualNetworkRules
      ipRules: networkAcls.?ipRules
  } : {
      bypass: 'AzureServices'
      defaultAction: 'Deny'
  }
  ```
- Make the UDT not nullable. If the parameter anyways has a default, making it's typo to nullable is an easy change.

File: avm/res/batch/batch-account/main.bicep
Comment (eriqua) [reply avm/res/batch/batch-account/main.bicep:77]: Never doubted that 😄 The problem is it did not reformat the whole module but only the main template and would have left us with a mixed up state. 
Thanks for approving the PR, now we are left with 5 missing formatting that I asked to be included as part of the open MR. We need to keep an eye on 2 PRs not yet opened. Storage account and flexible servers. You can resolve this one

Comment (AlexanderSehr) [reply avm/res/batch/batch-account/main.bicep:77]: You shall resolve it :P

Comment (eriqua) [reply avm/res/batch/batch-account/main.bicep:77]: And resolved it is

Comment (AlexanderSehr) [reply avm/res/batch/batch-account/main.bicep:77]: Much appreciated dear author :D

File: avm/res/storage/storage-account/README.md
Comment (eriqua) [reply avm/res/storage/storage-account/README.md:2055]: I thought you were commenting on the removed Default line. Otherwise I didn't get your comment

Comment (AlexanderSehr) [reply avm/res/storage/storage-account/README.md:2055]: I did @eriqua. The but while in the PR you linked defaults were updated, the above is a different change. And as per the comment below it was seemingly implemented as part of this PR to fix an issue. I'll resolve this comment as I only created it because I saw & commented the readme change before I saw the code change. As that one is the actual root cause, I suggest we only discuss the matter there.

File: avm/res/app-configuration/configuration-store/README.md
Comment (krbar) [reply avm/res/app-configuration/configuration-store/README.md:1]: Do I understand it correctly? The versions of module with (unrelated) issues should be reverted to their initial values, while versions of modules with successful pipelines should be left increased?

File: avm/res/storage/storage-account/main.bicep
Comment (krbar) [reply avm/res/storage/storage-account/main.bicep:72]: @fblix do you have any preference?

File: avm/res/app-configuration/configuration-store/README.md
Comment (AlexanderSehr) [reply avm/res/app-configuration/configuration-store/README.md:1]: I think that's what she is saying, yes. To be fair - if a workflow has not been publishing in a while, we did technically never increase the version. Then again - versions sometimes do make jumps when it comes to software releases ... 😄

Comment (eriqua) [reply avm/res/app-configuration/configuration-store/README.md:1]: What I'm suggesting is to either wait for the failing workflows to be fixed, or to remove the modules failing the validation from this PR. In the latter case we should track the removed ones with an issue, so that we don't miss updating them with this feature.
It would not make much difference whether we update minor explicitly (as it should be with semver) or just the patch.
If we merge failing modules now, we'll merge the code to main without releasing the corresponding version. We would pile up new features and fixes until the PR fixing the pipeline will allow all of them to be published.
Very annoying, I agree, but I guess it's a side effect of blocking non WAF compliant modules: any feature to those modules cannot be released until their validation is successful.

Comment (krbar) [reply avm/res/app-configuration/configuration-store/README.md:1]: It makes more sense to me to exclude some modules than to artificially manipulate the module versions. 

So let's agree on a following approach:
1. I would exclude following (failing) modules from this PR:
  - `app-configuration/configuration-store`
  - `container-registry/registry`
  - `databricks/workspace`
  - `event-hub/namespace`
  - `machine-learning-services/workspace`
  - `web/site`
2. I would also exclude the Storage Account module as it cannot be merged due to an open conversation regarding the `networkACLs` parameter
3. The I will create an issue for the "batch 2" of this PR and list all excluded modules to be handled there.

@eriqua @AlexanderSehr Fine for you?

Comment (AlexanderSehr) [reply avm/res/app-configuration/configuration-store/README.md:1]: I don't agree that we should not merge updates to modules that are not published for an unrelated reason. What's the worst that can happen? That the module is released with a bunch new features / fixes? 😄

On the other end, if we don't update them, we pile up work where somebody needs to manually rework certain modules after the fact. It adds to the effort big time. 👀

I may be a bit biased here as I only wait for this PR to work on two other bigger PRs to fix incorrect implementations. When I also have to wait for error-unrelated failing workflows to work in order to implement them, I won't implement them to begin with.

Comment (eriqua) [reply avm/res/app-configuration/configuration-store/README.md:1]: We agree we disagree :smile: except for the pickle.
I get the angle. I understand we are not losing features and fixes afterall, and that at some point they will all be published together. 
However I think we should avoid to consciously accept that. The intention of a PR should always be to publish to registry, whenever the PR qualifies for publishing of course. 
Deviating main from the published version could lead to confusion, but most importantly, it would add complexity in case of troubleshooting, backward tracking what was the change introducing an unexpected behavior.
Hence, I think our team efforts, although not fun at all, should go into fixing the failures first, and work on new features after. I'm not saying this PR should fix them, but the new features for failing modules should be put on hold.
This all due to a set of modules failing together all of a sudden, that we should probably avoid going forward.

File: avm/res/web/static-site/main.bicep
Comment (AlexanderSehr) [inline avm/res/web/static-site/main.bicep:247]: As per our conversation, please also update the `privateEndpoint.?manualPrivateLinkServiceConnections` assertions to `privateEndpoint.?isManualConnection`

Comment (krbar) [reply avm/res/web/static-site/main.bicep]: Done: https://github.com/Azure/bicep-registry-modules/pull/1472/commits/bafee36c2e2a9b9492f1a189d99672fe53822a70
Final test of all modules included in this PR in progress...

File: avm/res/app-configuration/configuration-store/README.md
Comment (krbar) [reply avm/res/app-configuration/configuration-store/README.md:1]: Excluded 6 failing modules + storage account:

- avm/res/app-configuration/configuration-store
- avm/res/container-registry/registry
- avm/res/databricks/workspace
- avm/res/event-hub/namespace
- avm/res/machine-learning-service/workspace
- avm/res/storage/storage-account
- avm/res/web/site

I created an issue to track these modules: https://github.com/Azure/bicep-registry-modules/issues/1583

File: avm/res/storage/storage-account/main.bicep
Comment (krbar) [reply avm/res/storage/storage-account/main.bicep]: @fblix I excluded the Storage Account from this module. Please update the PE implementation separately.

File: avm/res/cache/redis/main.bicep
Comment (AlexanderSehr) [inline avm/res/cache/redis/main.bicep:270]: I see that in quite a few files the `privateLinkServiceConnections` is incorrect. Per PE deployment there should be 2 checks for `isManualConnection` while some files have only one. I'll comment on each.

To fix, the privateLinkServiceConnections should be configured as
`privateLinkServiceConnections: privateEndpoint.?isManualConnection != true`

File: avm/res/cognitive-services/account/main.bicep
Comment (AlexanderSehr) [inline avm/res/cognitive-services/account/main.bicep:400]: See comment regarding [`privateLinkServiceConnections`](https://github.com/Azure/bicep-registry-modules/pull/1472#discussion_r1554834350).

File: avm/res/data-factory/factory/main.bicep
Comment (AlexanderSehr) [inline avm/res/data-factory/factory/main.bicep:310]: See comment regarding [`privateLinkServiceConnections`](https://github.com/Azure/bicep-registry-modules/pull/1472#discussion_r1554834350).

File: avm/res/desktop-virtualization/host-pool/main.bicep
Comment (AlexanderSehr) [inline avm/res/desktop-virtualization/host-pool/main.bicep:220]: See comment regarding [`privateLinkServiceConnections`](https://github.com/Azure/bicep-registry-modules/pull/1472#discussion_r1554834350).

File: avm/res/desktop-virtualization/workspace/main.bicep
Comment (AlexanderSehr) [inline avm/res/desktop-virtualization/workspace/main.bicep:114]: See comment regarding [`privateLinkServiceConnections`](https://github.com/Azure/bicep-registry-modules/pull/1472#discussion_r1554834350).

File: avm/res/event-grid/domain/main.bicep
Comment (AlexanderSehr) [inline avm/res/event-grid/domain/main.bicep:206]: See comment regarding [`privateLinkServiceConnections`](https://github.com/Azure/bicep-registry-modules/pull/1472#discussion_r1554834350).

File: avm/res/event-grid/topic/main.bicep
Comment (AlexanderSehr) [inline avm/res/event-grid/topic/main.bicep:216]: See comment regarding [`privateLinkServiceConnections`](https://github.com/Azure/bicep-registry-modules/pull/1472#discussion_r1554834350).

File: avm/res/insights/private-link-scope/main.bicep
Comment (AlexanderSehr) [inline avm/res/insights/private-link-scope/main.bicep:164]: See comment regarding [`privateLinkServiceConnections`](https://github.com/Azure/bicep-registry-modules/pull/1472#discussion_r1554834350).

File: avm/res/app-configuration/configuration-store/README.md
Comment (eriqua) [reply avm/res/app-configuration/configuration-store/README.md:1]: Meanwhile databricks workspace got green :)

Comment (krbar) [reply avm/res/app-configuration/configuration-store/README.md:1]: > Meanwhile databricks workspace got green :)

I will give it a second chance 😉

File: avm/res/cache/redis/main.bicep
Comment (krbar) [reply avm/res/cache/redis/main.bicep:272]: I re-created the Private Endpoints for all modules in this PR using the original specs, I think now everything should be fine.

File: avm/res/cognitive-services/account/main.bicep
Comment (krbar) [reply avm/res/cognitive-services/account/main.bicep:402]: Done, see above comment: https://github.com/Azure/bicep-registry-modules/pull/1472#discussion_r1554834350

File: avm/res/data-factory/factory/main.bicep
Comment (krbar) [reply avm/res/data-factory/factory/main.bicep:312]: Done, see above comment: https://github.com/Azure/bicep-registry-modules/pull/1472#discussion_r1554834350

File: avm/res/desktop-virtualization/host-pool/main.bicep
Comment (krbar) [reply avm/res/desktop-virtualization/host-pool/main.bicep:222]: Done, see above comment: https://github.com/Azure/bicep-registry-modules/pull/1472#discussion_r1554834350

File: avm/res/desktop-virtualization/workspace/main.bicep
Comment (krbar) [reply avm/res/desktop-virtualization/workspace/main.bicep:116]: Done, see above comment: https://github.com/Azure/bicep-registry-modules/pull/1472#discussion_r1554834350

File: avm/res/event-grid/domain/main.bicep
Comment (krbar) [reply avm/res/event-grid/domain/main.bicep:208]: Done, see above comment: https://github.com/Azure/bicep-registry-modules/pull/1472#discussion_r1554834350

File: avm/res/event-grid/topic/main.bicep
Comment (krbar) [reply avm/res/event-grid/topic/main.bicep:218]: Done, see above comment: https://github.com/Azure/bicep-registry-modules/pull/1472#discussion_r1554834350

File: avm/res/insights/private-link-scope/main.bicep
Comment (krbar) [reply avm/res/insights/private-link-scope/main.bicep:166]: Done, see above comment: https://github.com/Azure/bicep-registry-modules/pull/1472#discussion_r1554834350

File: avm/res/app-configuration/configuration-store/README.md
Comment (AlexanderSehr) [reply avm/res/app-configuration/configuration-store/README.md:1]: Looks like an issue with the `-managed` rg that is auto-created but not auto-removed. I'd suggest to cleanup anything that looks like Databricks in the Subscription before re-running the failed jobs

**Review by AlexanderSehr — DISMISSED** [review]
Content-wise approved

File: avm/res/app-configuration/configuration-store/README.md
Comment (krbar) [reply avm/res/app-configuration/configuration-store/README.md:1]: > Looks like an issue with the `-managed` rg that is auto-created but not auto-removed. I'd suggest to cleanup anything that looks like Databricks in the Subscription before re-running the failed jobs

Not sure about this. I tried in two different subscriptions, each time changing to a new, unused Name Prefix. I think we should throw Databricks out of this PR, what do you think?

Comment (AlexanderSehr) [reply avm/res/app-configuration/configuration-store/README.md:1]: If that means we can get the PR in @krbar , sure :)

Comment (krbar) [reply avm/res/app-configuration/configuration-store/README.md:1]: Databricks is out. Sorry @eriqua, I really tried 😅...

Comment (eriqua) [reply avm/res/app-configuration/configuration-store/README.md:1]: Fine for me

**Review by AlexanderSehr — APPROVED** [review]

File: avm/res/app-configuration/configuration-store/README.md
Comment (AlexanderSehr) [reply avm/res/app-configuration/configuration-store/README.md:1]: Resolving to unblock

**Pull request merged**
